### DISCUSSION TITLE
Logging data from sensor to SD card

### DISCUSSION THREAD

**Discussion by Arthos2** [arendst/Tasmota#18056] [Q&A] [answered]
I'm working on a project to log data from a smart meter to an SD card. The data is received via IR sensor which is connected to an ESP32 DevKit v4 on which Tasmota 12.3.1 is installed. The sensor is already working and Tasmota is printing the sensor data on the screen. I also was able to connect the SD card module to the ESP32. The filesystem seems to work as I can see it in Tasmota and can create and delete files. However, when I enable my script the console repeatedly states that it fails to open the file I'm referring to ("file open failed"). I don't know how to proceed and where I could have made a mistake. Could anyone please help?

What I have already tried:
- including all #define in my .bin which may be necessary for the SD card module and scripting from discussion [#13938](https://github.com/arendst/Tasmota/discussions/13938#discussion-3726437) (i know that it was for the ESP8266, but I could not find any other reference)
- disconnecting the SD card module and refering to a file on the ESP's internal filesystem

SD card is connected to the following pins:
GPIO5 = SDCard CS
GPIO18 = SPI CLK
GPIO19 = SPI MISO
GPIO23 = SPI MOSI

My script looks as follows:

`>D
v1=0
fr=0
res=0
str=""

'>B
TelePeriod 30
=>sensor53 r

'>M 1
+1,1,o,0,300,e350,3,30,2F3F210D0A,063030300D0A

1,1.8.1(@1,Gesamtverbrauch,KWh,total,3
1,31.7(@1,Strom aktuell,A,current,3)
1,32.7(@1,Spannung,V,voltage,0)

'>T
v1=sml[1]
fr=fo("slog.txt" 1)
str=s(tstamp)+"\t"+s(v1)+"\n"
res=fw(str fr)
fc(fr)

#`

**Comment by Jason2866** [comment, marked as answer]
take a look at #17955 or #16074

**Reply by Arthos2** [reply]
Thank's for the suggestions. Unfortunately, the discussions were not very helpful. After a few more hours I finally found the solution, which is very simple. I just have to put a "/" before the file name. Therefore, the file open command looks as follows:
`fr=fo("/slog.txt" 1)`

Moreover, I have only used the following defines in user_config_override.h to get the SD card, file system and sensor working:
```
// standard defines to use scripting language
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

//define file system and SD card support
#ifndef USE_UFILESYS 
#define USE_UFILESYS
#endif
#ifndef USE_SDCARD
#define USE_SDCARD
#endif

//defines for SML use (smart meter)
#ifndef USE_SML_M
#define USE_SML_M
#endif
```
